Add Support for Writing Saturation Function Tables to INIT File

bska opened the pull request:
This change-set extends the INIT file Table writing to also output the saturation function tables in normalised form.  We support generating the INIT file's SGFN, SOFN, and SWFN tables&mdash;including the derivative information&mdash;for both two and three phases from both families of saturation function keywords (S{G,W}OF and S{G,W}FN + SOF{2,3}).  We do not yet support generating the appropriate tables from input keyword SLGOF.

The main part of this change-set is commit ~~35eac9f~~ 07aad01 which adds the table normalisation and padding for all combinations of supported saturation function keywords and a collection of unit tests to verify that we produce the tables we expect.  This commit builds on commit ~~3a3738c~~ 81d64c7 which introduces a new helper class to simplify indexing into the linearised and padded "TAB" vector entries and enables using standard algorithms such as `std::copy()` and `std::transform()` to assign the appropriate column values rather than using explicit looping.  This commit also adds a helper function,
```C++
Opm::DifferentiateOutputTable::calcSlopes()
```
that is useful as a means of centralising the calculation of derivatives of the piecewise linear interpolants that are implicitly specified through the normalised saturation function tables.  Derivative/slope information must accompany the primary output columns according to the file format protocol.

As part of creating the unit tests I have elected to remove the member function
```C++
void Opm::Tables::fwrite(ERT::FortIO& fortio) const
```
in favour of providing equivalent functionality through a new free function,
```C++
void Opm::fwrite(const Tables& tables, ERT::FortIO& fortio)
```
that is implemented in terms of acquiring read-only references to the internal representations of the "TABDIMS" and "TAB" vectors (see commit ~~346d977~~ c3cb948) and forming `ERT::EclKW<>` instances based on those data values.

Do note that as this change-set explicitly modifies the "INIT" file, we will break any regression test that compares the contents of that file before and after this change.  We should therefore only install this if we're comfortable updating the relevant baseline files.

Finally, a few notes about the implementation:

1. We create scoped helper functions to handle each saturation function table in the appropriate context, for instance
    ```C++
    std::vector<double>
    SatFunc::Oil::TwoPhase::fromSGOF(const std::size_t          numRows,
                                     const Opm::TableContainer& sgof);
    ```
    to create the two-phase "SOFN" vector entries from information stored in tables entered with the `SGOF` keyword.

2. Each of these scoped helper functions are, in turn, implemented in terms of a helper template function
    ```C++
    template <class BuildDependent>
    std::vector<double>
    createSatfuncTable(const std::size_t numTab,
                       const std::size_t numRows,
                       const std::size_t numDep,
                       BuildDependent&&  buildDeps)
    ```
    that implements a simple loop over saturation function regions and adds appropriate derivative information to each normalised output table.  The callable entity `buildDeps`&mdash;not necessarily a good name&mdash;is responsible for assigning independent variables and the table-required dependent variables (e.g., relative permeability for water and oil-water capillary pressure) for a single table/saturation region and is invoked on each iteration of `createSatfuncTable<>()`.

3. Creating the three-phase "SOFN" table from family one (SGOF, SWOF) keywords is slightly involved.  We need to merge the tables based on common (and unique) values of the oil saturation which tends to expand the number of rows needed to represent the table.  This expansion in turns means that missing relative permeability values must be defined through piecewise linear interpolation in the "KrOX" column of the pertinent input table.  We defer the task of creating a unique sequence of merge oil saturation values to the standard library routine `std::set_union()` and then only need to create a wrapper class to facilitate oil saturation look-up and relative permeability value either in terms of a saturation node index or a saturation value.

bska committed:
Table Output: Switch to One-Based Offset Indices

This is needed for compatibility with Fortran and external software
that reads the .INIT file.

bska committed:
Table Output: Add Facility to Simplify Indexing Details

This commit introduces a new helper class, LinearisedOutputTable,
that is oriented towards managing tabular data such as the PVT or
saturation functions in an ECL result set (typically, .INIT file).
The class knows about row padding and the column oriented nature of
Fortran-like tables.

While here, and in anticipation of adding output of tabulated
saturation functions, also provide a mechanism for calculating
slopes of piecewise linear interpolants.  The ECL format stipulates
that function derivatives be output along with the actual table
data for most functions.

bska committed:
Linearised Output Table: Store Derivatives at Right-Hand End Point

This is more in keeping with the intended application--at least for
saturation function output to the INIT result set.

While here, also remove an unused data table (SWOF from SPE1).

bska committed:
Table Output: Don't Store ECL Keywords Internally

This commit switches the internal implementation of the 'tab' and
'tabdims' vectors from being a mix of ECL keyword structures and
std::vector<>s to being purely std::vector<>s.  We remove the
fwrite() member function and rather reimplement the feature in terms
of new query member functions

    const std::vector<int>& tabdims() const
    const std::vector<double>& tab() const

that return read-only references to internal data members.

bska committed:
INIT File: Add Support for Writing Saturation Function Tables

This commit extends the INIT file Table writing to also output the
saturation function tables in normalised form.  We support
generating the INIT file's SGFN, SOFN, and SWFN tables--including
the derivative information--for both two and three phases from both
families of saturation function keywords (S{G,W}OF and S{G,W}FN +
SOF{2,3}).  We do not yet support generating the appropriate tables
from input keyword SLGOF.

We leverage the LinearisedOutputTable to abstract away some of the
details of indexing into the linear 'tab' array and create a helper
function to loop across saturation function regions whence each
table generator need only be concerned about the specific data
pertaining to that region's saturation function (i.e., the values of
independent and dependent variates and number of active rows).  We
allocate the output tables according to information in the TABDIMS
keyword, notably the number of saturation nodes (item 3 of TABDIMS).

Generating the three-phase SOFN table, which has the same columns as
the input table SOF3, in the case of family One (SGOF and SWOF) is
slightly involved.  In particular we need to join the SGOF and SWOF
tables on common oil saturation values and derive/insert missing
KroX data through piecewise linear interpolation in the appropriate
input table.  We defer the details of merging on common (and uniqe)
oil saturation values to the standard library routine 'set_union()'
which outputs repeated saturation values exactly once rather than
for each occurrence.  That way we only need to wrap the input tables
in a simple class to facilitate look-up of oil saturation values and
the mechanics of piecewise linear interpolation of the relative
permeability column for oil.  Due to this merging process, the
number of rows in SOFN is twice the number of saturation nodes in
this case.

Since this support depends on being able to determine which phases
are active in a particular run, what table sizes to use for the
output and which keywords are being used to represent the saturation
function tables we introduce a single new public member function

    void Tables::addSatFunc(const EclipseState& es)

This function determines the run's active phases and dispatches to
the new helper functions

    void Tables::addSatFunc_FamilyOne()
    void Tables::addSatFunc_FamilyTwo()

according to which keyword family exists in the input deck.  These
functions in turn call additional helper functions to handle each
phase table separately.  For instance addSatFunc_FamilyOne() uses
the helper function

    SatFunc::Oil::ThreePhase::fromSGOFandSWOF()

to generate a three-phase SOFN table based on the information in the
SGOF and SWOF tables.

Information about slopes/derivatives for the piecewise interpolants
of the dependent variates is computed through the helper function

    DifferentiateOutputTable::calcSlopes()

whence the first entry of each derivative column is undefined (i.e.,
assigned the sentinel value 1e+20).  This should arguably be zero
instead.  Further testing and comparison with INIT result sets
generated by ECL will inform the decision here.

We add unit tests for all combinations of number of active phases
(two or three) and which saturation function keyword family is
present in the simulation case.  The unit tests use the tables from
SPE 9 for keyword family One and the tables from SPE 1 for keyword
family Two.  Comparison data is extracted directly from the ECL INIT
file in the case of the SPE 9 tables and derived from independent
calculation for the SPE 1 tables.

bska committed:
Init File: Hook Saturation Function Writing Up to System

joakim-hove commented:
This was a mighty piece of work; will probably need some to review (i.e. understand ...) it. Epic commit messages :-)

joakim-hove commented on opm/output/eclipse/LinearisedOutputTable.hpp:
```diff
@@ -0,0 +1,168 @@
+
+    private:
+        /// Internal buffer for tabular data.
+        std::vector<double> data_;
```
I'm sorry to bother you with this - but please please; those trailing underscores hurt my eyes - there is no tradition for using them in the rest of opm-output.

bska replied:
I can probably name it `m_*` if that's better, but I can't keep the existing name as there can't be two class-scope identifiers called `tabdims`.

bska commented:
> This was a mighty piece of work

Thanks.  I've been building it on and off for the past month or so.  Many iterations...

> will probably need some to review (i.e. understand ...)

There's no rush.  Feel free to ask questions about it.

joakim-hove commented:
> Feel free to ask questions about it.

I guess I will - mainly for my own interest and curiosity; are you on Texas time or Gaustad time?

joakim-hove commented:
> [... ] keywords is slightly involved. 

I recognize an understatement when I see one.

bska commented:
> are you on Texas time or Gaustad time?

Gaustad.

>> [... ] keywords is slightly involved.
> 
> I recognize an understatement when I see one.

In that case, take comfort in the knowledge that it would be much more complicated if `std::set_union()` did not exist.

bska committed:
Table Output: Switch Naming Convention for Data Members

Mostly to remove trailing underscores.  The internal 'tabdims'
vector in Opm::Tables has been renamed to 'm_tabdims' to avoid
conflicting with the member function named 'tabdims'.

bska commented:
I have rebased this change-set onto current master and addressed a few issues that presented during review.  I now consider this to be the more-or-less final version of the work and would like to submit it to final approval.

joakim-hove commented:
jenkins build this with downstreams update_data please

joakim-hove commented:
jenkins build this with downstreams opm-data=279 please
